### ISSUE TITLE
Difference of parameters between code and paper

### ISSUE THREAD

**Issue by SungwanRyu91** [arpanpoudel/lenslessimaging#2]
Hello. Can I ask you about why does the number of parameters is different with number that reported in the paper?
I checked the number of parameters in [Reconstruction_demo.ipynb] model(model=EnsembleModel(admm,model_unet)) with fvcore library.
Below is the code and the result.
```python
import torch
import torch.nn as nn
from fvcore.nn import FlopCountAnalysis, flop_count_table

admm = model_builder.build_model(batch_size=1,iterations=5,device=my_device)
model_unet=Unet(dim=36,channels=3,dim_mults=(1,2,4,8)).to(my_device)
model=EnsembleModel(admm,model_unet)

def count_flops(model, input_shape):
    # Create a dummy input with the given shape
    dummy_input = torch.zeros(input_shape)

    # Calculate FLOPs using FlopCountAnalysis
    flops = FlopCountAnalysis(model, dummy_input)
    
    # Generate a detailed FLOPs table
    flops_table = flop_count_table(flops, max_depth=4)

    return flops_table, flops.total()

# Replace this with your actual model
 

# Replace (1, 3, 224, 224) with the correct input shape for your model
input_shape = (1, 3, 270, 480)
flops_table, total_flops = count_flops(model, input_shape)

print(flops_table)
print(f"Total FLOPs: {total_flops:.2e}")

``` 
```shell
| module                                      | #parameters or shape   | #flops      |
|:--------------------------------------------|:-----------------------|:------------|
| model                                       | 18.793M                | 91.997G     |
|  admm_model                                 |  0.648M                |  0.389M     |
|   admm_model.mu1                            |   (5,)                 |             |
|   admm_model.mu2                            |   (5,)                 |             |
|   admm_model.mu3                            |   (5,)                 |             |
|   admm_model.tau                            |   (5,)                 |             |
|   admm_model.h_var                          |   (270, 480)           |             |
|   admm_model.LtL                            |   (540, 960)           |             |
|  denoise_model                              |  18.145M               |  91.997G    |
|   denoise_model.init_conv                   |   3.552K               |   0.457G    |
|    denoise_model.init_conv.weight           |    (24, 3, 7, 7)       |             |
|    denoise_model.init_conv.bias             |    (24,)               |             |
|   denoise_model.downs                       |   7.765M               |   51.762G   |
|    denoise_model.downs.0                    |    0.129M              |    15.88G   |
|     denoise_model.downs.0.0                 |     41.28K             |     5.365G  |
|     denoise_model.downs.0.1                 |     48.78K             |     6.345G  |
|     denoise_model.downs.0.2.fn              |     18.612K            |     3.497G  |
|     denoise_model.downs.0.3                 |     20.772K            |     0.672G  |
|    denoise_model.downs.1                    |    0.456M              |    13.051G  |
|     denoise_model.downs.1.0                 |     0.145M             |     4.705G  |
|     denoise_model.downs.1.1                 |     0.191M             |     6.196G  |
|     denoise_model.downs.1.2.fn              |     37.224K            |     1.483G  |
|     denoise_model.downs.1.3                 |     83.016K            |     0.667G  |
|    denoise_model.downs.2                    |    1.736M              |    12.031G  |
|     denoise_model.downs.2.0                 |     0.575M             |     4.628G  |
|     denoise_model.downs.2.1                 |     0.755M             |     6.076G  |
|     denoise_model.downs.2.2.fn              |     74.448K            |     0.67G   |
|     denoise_model.downs.2.3                 |     0.332M             |     0.657G  |
|    denoise_model.downs.3                    |    5.443M              |    10.8G    |
|     denoise_model.downs.3.0                 |     2.291M             |     4.537G  |
|     denoise_model.downs.3.1                 |     3.003M             |     5.949G  |
|     denoise_model.downs.3.2.fn              |     0.149M             |     0.314G  |
|   denoise_model.ups                         |   4.173M               |   20.206G   |
|    denoise_model.ups.0                      |    3.142M              |    6.242G   |
|     denoise_model.ups.0.0                   |     1.98M              |     3.924G  |
|     denoise_model.ups.0.1                   |     0.755M             |     1.496G  |
|     denoise_model.ups.0.2.fn                |     74.448K            |     0.165G  |
|     denoise_model.ups.0.3                   |     0.332M             |     0.657G  |
|    denoise_model.ups.1                      |    0.814M              |    6.621G   |
|     denoise_model.ups.1.0                   |     0.503M             |     4.049G  |
|     denoise_model.ups.1.1                   |     0.191M             |     1.538G  |
|     denoise_model.ups.1.2.fn                |     37.224K            |     0.368G  |
|     denoise_model.ups.1.3                   |     83.016K            |     0.667G  |
|    denoise_model.ups.2                      |    0.218M              |    7.343G   |
|     denoise_model.ups.2.0                   |     0.13M              |     4.211G  |
|     denoise_model.ups.2.1                   |     48.78K             |     1.586G  |
|     denoise_model.ups.2.2.fn                |     18.612K            |     0.874G  |
|     denoise_model.ups.2.3                   |     20.772K            |     0.672G  |
|   denoise_model.mid_block1                  |   3.003M               |   5.949G    |
|    denoise_model.mid_block1.ds_conv         |    14.4K               |    27.942M  |
|     denoise_model.mid_block1.ds_conv.weight |     (288, 1, 7, 7)     |             |
|     denoise_model.mid_block1.ds_conv.bias   |     (288,)             |             |
|    denoise_model.mid_block1.net             |    2.989M              |    5.921G   |
|     denoise_model.mid_block1.net.0          |     0.576K             |     2.851M  |
|     denoise_model.mid_block1.net.1          |     1.494M             |     2.956G  |
|     denoise_model.mid_block1.net.3          |     1.152K             |     5.702M  |
|     denoise_model.mid_block1.net.4          |     1.493M             |     2.956G  |
|   denoise_model.mid_attn.fn                 |   0.148M               |   1.299G    |
|    denoise_model.mid_attn.fn.fn             |    0.148M              |    1.296G   |
|     denoise_model.mid_attn.fn.fn.to_qkv     |     0.111M             |     0.219G  |
|     denoise_model.mid_attn.fn.fn.to_out     |     37.152K            |     72.991M |
|    denoise_model.mid_attn.fn.norm           |    0.576K              |    2.851M   |
|     denoise_model.mid_attn.fn.norm.weight   |     (288,)             |             |
|     denoise_model.mid_attn.fn.norm.bias     |     (288,)             |             |
|   denoise_model.mid_block2                  |   3.003M               |   5.949G    |
|    denoise_model.mid_block2.ds_conv         |    14.4K               |    27.942M  |
|     denoise_model.mid_block2.ds_conv.weight |     (288, 1, 7, 7)     |             |
|     denoise_model.mid_block2.ds_conv.bias   |     (288,)             |             |
|    denoise_model.mid_block2.net             |    2.989M              |    5.921G   |
|     denoise_model.mid_block2.net.0          |     0.576K             |     2.851M  |
|     denoise_model.mid_block2.net.1          |     1.494M             |     2.956G  |
|     denoise_model.mid_block2.net.3          |     1.152K             |     5.702M  |
|     denoise_model.mid_block2.net.4          |     1.493M             |     2.956G  |
|   denoise_model.final_conv                  |   48.891K              |   6.359G    |
|    denoise_model.final_conv.0               |    48.78K              |    6.345G   |
|     denoise_model.final_conv.0.ds_conv      |     1.8K               |     0.229G  |
|     denoise_model.final_conv.0.net          |     46.98K             |     6.117G  |
|    denoise_model.final_conv.1               |    0.111K              |    13.997M  |
|     denoise_model.final_conv.1.weight       |     (3, 36, 1, 1)      |             |
|     denoise_model.final_conv.1.bias         |     (3,)               |             |
Total FLOPs: 9.20e+10

```

In the [paper](https://openaccess.thecvf.com/content/WACV2024W/VAQ/papers/Poudel_DeepLIR_Attention-Based_Approach_for_Mask-Based_Lensless_Image_Reconstruction_WACVW_2024_paper.pdf), the model's parameter is 19.3M.
> This model consists of 19.3 million parameters with 18.3 million trainable parameters. (in 4.1. Experminetal Setup)

Can I ask you about the code setup(to call 'model_builder.build_model', 'Unet' class) to reproduce DeepLIR model in paper?
Thanks for read my issue.

**Comment by Phaaarus** [comment]
I resolved this issue by exchanging emails with arpanpoudel.
Below is the full text of the emails we exchanged.
---

# Ask about issue + performance of pre-trained weight on github repo + the method DeepLIR used to address the issue with the test dataset

> I have thoroughly read your paper. However, I have some questions regarding the content of your paper.
> First thing is that I uploaded on the [issues] in github(https://github.com/arpanpoudel/lenslessimaging). The number of parameters is different from the number reported in the paper. With the parameters of pre-trained weights you linked at github repository(https://drive.google.com/file/d/1IMMPp7pR7XcUdbayVngp3HeI3LdT0rsf/view), there are 18.793M parameters. But in the DeepLIR paper, the reported model's parameter is 19.3M. (This model consists of 19.3 million parameters with 18.3 million trainable parameters. (in 4.1. Experminetal Setup) Could you clarify this difference for me?
> Secondly I checked the metric scores (MSE, PSNR, LPIPS) of pre-trained weights with DiffuserCam Lensless Mirflickr Dataset by "result_test.py" code in your github repo. But there are some differences with results and reported scores. I got the scores: avg mse: 0.00260708 avg psnr: 26.624216 avg lpips: 0.14077479. But the reported scores are MSE: 0.0022, PSNR: 26.40, LPIPS: 0.1412. Can you explain these differences?
> Third, I noticed that the "im1.jpg.tiff" file does not exist within the dataset, but it is mentioned in the dataset_test.csv file. It makes error when I test the model. So I just delete im1.jpg.tiff text in dataset_test.csv file. How did you address this discrepancy?

# Answer from Arpanpoudel

> Thank you for reaching out. I couldn't find out the exact parameters for the model to reproduce from the paper. However, I had used the torchinfo.summary to calculate total number of parameters and trainable parameters.
> 
> Here is what I reproduced:
> Total params: 18,792,791 Trainable params: 18,144,791 Non-trainable params: 648,000.
>
> If I remember correctly, I created the reconstruction_demo.ipynb a few months after the paper had been published and the model would slightly defer due to the discrepancy in the parameters used for the results obtained during the paper report. I apologize for the difference. 
> 
> Regarding the evaluation metrics, the results will slightly alter depending on the code you use to calculate the metrics and the number of trails.  I see your replication does not vary much from the reported score (by two place decimals) and I assume it is due to the code and number of trials for the evaluation metrics. 
> 
> Lastly, the dataset belongs to Kristina's lab and not mine. I had noticed the issue and contacted the author but did not get much. Therefore, I had to remove those files. You may follow that approach. I hope this is helpful for you.

---
I hope this Q&A can resolve your questions.

**Comment by SungwanRyu91** [comment]
I resolved this issue by exchanging emails with arpanpoudel.
Below is the full text of the emails we exchanged.
---

# Ask about issue + performance of pre-trained weight on github repo + the method DeepLIR used to address the issue with the test dataset

> I have thoroughly read your paper. However, I have some questions regarding the content of your paper.
> First thing is that I uploaded on the [issues] in github(https://github.com/arpanpoudel/lenslessimaging). The number of parameters is different from the number reported in the paper. With the parameters of pre-trained weights you linked at github repository(https://drive.google.com/file/d/1IMMPp7pR7XcUdbayVngp3HeI3LdT0rsf/view), there are 18.793M parameters. But in the DeepLIR paper, the reported model's parameter is 19.3M. (This model consists of 19.3 million parameters with 18.3 million trainable parameters. (in 4.1. Experminetal Setup) Could you clarify this difference for me?
> Secondly I checked the metric scores (MSE, PSNR, LPIPS) of pre-trained weights with DiffuserCam Lensless Mirflickr Dataset by "result_test.py" code in your github repo. But there are some differences with results and reported scores. I got the scores: avg mse: 0.00260708 avg psnr: 26.624216 avg lpips: 0.14077479. But the reported scores are MSE: 0.0022, PSNR: 26.40, LPIPS: 0.1412. Can you explain these differences?
> Third, I noticed that the "im1.jpg.tiff" file does not exist within the dataset, but it is mentioned in the dataset_test.csv file. It makes error when I test the model. So I just delete im1.jpg.tiff text in dataset_test.csv file. How did you address this discrepancy?

# Answer from Arpanpoudel

> Thank you for reaching out. I couldn't find out the exact parameters for the model to reproduce from the paper. However, I had used the torchinfo.summary to calculate total number of parameters and trainable parameters.
> 
> Here is what I reproduced:
> Total params: 18,792,791 Trainable params: 18,144,791 Non-trainable params: 648,000.
>
> If I remember correctly, I created the reconstruction_demo.ipynb a few months after the paper had been published and the model would slightly defer due to the discrepancy in the parameters used for the results obtained during the paper report. I apologize for the difference. 
> 
> Regarding the evaluation metrics, the results will slightly alter depending on the code you use to calculate the metrics and the number of trails.  I see your replication does not vary much from the reported score (by two place decimals) and I assume it is due to the code and number of trials for the evaluation metrics. 
> 
> Lastly, the dataset belongs to Kristina's lab and not mine. I had noticed the issue and contacted the author but did not get much. Therefore, I had to remove those files. You may follow that approach. I hope this is helpful for you.

---
I hope this Q&A can resolve your questions.
